### ISSUE TITLE
[Core feature] Create or document a fast way to query the status of large workflows

### ISSUE THREAD

**Issue by Tom-Newton** [flyteorg/flyte#4056]
### Motivation: Why do you think this is important?

For large workflows with ~1000s of nodes its difficult to monitor the overall progress. At this scale we need to be able to get things like counts of how many nodes or sub-workflows are in each status. e.g. 900 success, 200: unknown 2 failed, 98: in progress.

I think supporting large workflows like this would be a valuable feature and its critical to what I want to do with flyte. 

### Goal: What should the final outcome look like, ideally?

I think ideal would be an interface similar to `remote.sync_execution(sync_nodes=True)` but fast for large workflows. This would be very flexible. 

Other interfaces would also be fine - I'm mostly just interested in it being fast. 

### Describe alternatives you've considered 

The flyte UI:
It displays lists and graphs but at the scale of 1000s of nodes these are impossible to parse by eye. Additionally in tends to crash my browser. 

`flytectl get execution`
Can get information about nodes when using `--details` but it seems to be incomplete. Writing to a `.yaml` file and searching, I find quite a lot of nodes are missing. 

`flytekit` `remote.sync_execution(sync_nodes=True)`
This does fetch all the important information and could certainly be parsed by some python code to extra whatever metrics are needed. The problem is that it takes about 12 minutes to run on a workflow with 3000 nodes.
EDIT: It actually doesn't fetch information about nodes that haven't started processing yet. So nodes that would show with unknown status on the UI are missed.

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

**Comment by kumare3** [comment]
cc @pingsutw when we added system tags did we add a query method for this - in list executions?

**Comment by pingsutw** [comment]
yup. List with filter: http://localhost:30082/api/v1/executions/flytesnacks/development?filters=value_in(admin_tag.name,flyte%3Btest)&limit=25&sort_by.direction=DESCENDING&sort_by.key=created_at

https://github.com/flyteorg/flyte/issues/3960

**Comment by Tom-Newton** [comment]
Thanks for the replies but I'm afraid I don't really understand how the tags can help. I want to see the status of all the nodes in a particular workflow quickly but I can't find any API that can list all the nodes in a workflow. All the APIs seem to list only immediate child nodes, rather that the whole graph. 

I don't really think filtering is the problem, all the APIs I've tested already have too much filtering.

**Comment by Tom-Newton** [comment]
I would still be keen to know if there is a better way but I ended up creating a version of `remote.sync_execution(sync_nodes=True)` that make gRPC requests in many parallel threads by wrapping `iterate_node_executions` and running it with `asyncio`. Now the 12 minutes is now more like 10 seconds.

**Comment by EngHabu** [comment]
This is awesome, @Tom-Newton...
flytekit (particularly remote) needs asyncio everywhere... Would you be open to redoing your change within flytekit codebase to support asyncIO? even if we do it one API at a time (or create a separate asyncRemote type with smaller interface initially until we migrate all APIs)

**Comment by Tom-Newton** [comment]
What I did was a fairly naive approach that could definitely be optimised. The best way though would definitely be to add native support for `asyncio` in `flytekit` as you say. Anything that uses gRPC could probably benefit. I'm a total noob when it comes to `asyncio` and to gRPC but it looks like we could use https://grpc.github.io/grpc/python/grpc_asyncio.html. Potentially I could contribute to something like this but probably not in the near term.

**Comment by Tom-Newton** [comment]
If anyone is interested in the code I'm using personally 
[flyte_remote.py.txt](https://github.com/flyteorg/flyte/files/12732200/flyte_remote.py.txt) (Its got a `.txt` extension because github doesn't allow uploading `.py`)
